File: strategies/ot/random_clicks.js
/**
 * random_clicks.js — Random baseline strategy for /sphere trace (ot).
 *
 * Picks a random unclicked cell on every turn.
 * No constraint inference — does not use ship geometry.
 *
 * This is the simplest possible strategy — no state, no inference.
 * See stateful.js (oq/) for per-game state usage, global_state.js (oc/)
 * for cross-game global state usage.
 */

"use strict";

const { OTStrategy, register } = require("../../interface/strategy.js");

class RandomOTStrategy extends OTStrategy {
  nextClick(board, meta, gameState) {
    const clicked = new Set(board.filter(c => c.clicked).map(c => c.row * 5 + c.col));
    const unclicked = [];
    for (let r = 0; r < 5; r++)
      for (let c = 0; c < 5; c++)
        if (!clicked.has(r * 5 + c)) unclicked.push([r, c]);
    if (unclicked.length === 0) return { row: 0, col: 0, gameState };
    const [row, col] = unclicked[Math.floor(Math.random() * unclicked.length)];
    return { row, col, gameState };
  }
}

register(new RandomOTStrategy());

File: strategies/ot/random_clicks.py
"""Random baseline strategy for /sphere trace (ot).

Picks a random unclicked cell on every turn.
No constraint inference — does not use ship geometry.

This is the simplest possible strategy — no state, no inference.
See stateful.py (oq/) for per-game state usage, global_state.py (oc/)
for cross-game global state usage.
"""

import random
from typing import Any

from interface.strategy import OTStrategy


class RandomOTStrategy(OTStrategy):
    def next_click(
        self,
        board: list[dict[str, Any]],
        meta: dict[str, Any],
        game_state: Any,
    ) -> tuple[int, int, Any]:
        clicked = {(c["row"], c["col"]) for c in board if c["clicked"]}
        unclicked = [
            (r, c)
            for r in range(5)
            for c in range(5)
            if (r, c) not in clicked
        ]
        if not unclicked:
            return 0, 0, game_state
        row, col = random.choice(unclicked)
        return row, col, game_state  # game_state is None — stateless strategy
